tests/unittest/test_gcs_crashstorage.py

```diff
@@ -0,0 +1,137 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import io
+from unittest.mock import call, Mock, patch
+
+import pytest
+from google.cloud.exceptions import NotFound, Unauthorized
+
+from testlib.mini_poster import multipart_encode
+
+
+@pytest.fixture(autouse=True)
+def mock_generate_test_filepath():
+    with patch("antenna.ext.gcs.crashstorage.generate_test_filepath") as gtfp:
+        gtfp.return_value = "test/testwrite.txt"
+        yield
+
+
+class TestGcsCrashStorageIntegration:
+    logging_names = ["antenna"]
+
+    @pytest.mark.parametrize("bucket_name", ["fakebucket", "fakebucket.with.periods"])
+    @patch("google.cloud.storage.Client")
+    def test_crash_storage(self, MockStorageClient, client, bucket_name):
+        crash_id = "de1bb258-cbbf-4589-a673-34f800160918"
+        data, headers = multipart_encode(
+            {
+                "uuid": crash_id,
+                "ProductName": "Firefox",
+                "Version": "1.0",
+                "upload_file_minidump": ("fakecrash.dump", io.BytesIO(b"abcd1234")),
+            }
+        )
+
+        # Rebuild the app the test client is using with relevant configuration.
+        client.rebuild_app(
+            {
+                "CRASHMOVER_CRASHSTORAGE_CLASS": "antenna.ext.gcs.crashstorage.GcsCrashStorage",
+                "CRASHMOVER_CRASHSTORAGE_BUCKET_NAME": bucket_name,
+            }
+        )
+
+        result = client.simulate_post("/submit", headers=headers, body=data)
+
+        # Verify the collector returns a 200 status code and the crash id
+        # we fed it.
+        assert result.status_code == 200
+        assert result.content == f"CrashID=bp-{crash_id}\n".encode("utf-8")
+
+        # Assert we uploaded files to gcs
+        mock_client = MockStorageClient.return_value
+        bucket = mock_client.get_bucket.return_value
+        blob = bucket.blob.return_value
+        upload_calls = blob.upload_from_string.mock_calls
+        blob_calls = [c for c in bucket.blob.mock_calls if c not in upload_calls]
+        bucket_calls = [
+            c
+            for c in mock_client.get_bucket.mock_calls
+            if c not in blob_calls and c not in upload_calls
+        ]
```

Comment:
At some point, I changed how I test things to use the local dev environment emulators rather than mocking everything out. It's much safer in the long run and not much harder except when one is testing exceptional circumstances like permissions.

I think we should do these tests more like what we have in Socorro:

https://github.com/mozilla-services/socorro/blob/main/socorro/tests/external/boto/test_crashstorage.py

Is that doable or is it tricky because we have both S3 and GCS stuff?

Reply:
shouldn't be tricky at all, will do